### ISSUE TITLE
Support For Columstore Index Storage

### ISSUE THREAD

**Issue by StrangeWill** [aspnet/EntityFrameworkCore#15733]
Currently it appears the only way I can roll Columnstore Indexes (https://docs.microsoft.com/en-us/sql/relational-databases/indexes/columnstore-indexes-overview?view=sql-server-2017) is to basically write a ton of SQL during a migration and leave the database in a state that the model is not really aware about.

Is there any planned support for these unique types of indexes for the SQL Server provider of EF Core?

**Comment by ErikEJ** [comment]
@StrangeWill You could also use SSDT for database first development

**Comment by StrangeWill** [comment]
Hmm, yeah I guess going back to SSDT is an option, code-first and EF migrations brought us a _ton_ of benefits though.

**Comment by ErikEJ** [comment]
EF Core Power Tools can generate DbContext and POCO classes from your SSDT project...

**Comment by StrangeWill** [comment]
> EF Core Power Tools can generate DbContext and POCO classes from your SSDT project...

Yep, it just requires us to do double up/change a lot of our process:

- Application cannot manage it's own database state, we need to add new processes both for dev (create new SSDT changes, re-import new context/POCOs), and our CI process (removing our ability to build on Linux).
- SSDT tools now need to be deployed on everyone's machines (we have some devs that are 100% Linux now being as that was a huge plus to Core and SQL server going Linux).

We'll consider that but it's definitely has a non-zero cost, we'd probably rather just hack it into the migration over lose all the benefits we get with code-first migrations.
-

**Comment by ErikEJ** [comment]
Fwiw, you can run sqlpackage on Linux

**Comment by tekrarCo** [comment]
Is code first EF Core has plan to support Columstore or not? how many years EF Core wants to be behind ?

**Comment by StrangeWill** [comment]
We just threw some SQL shenanigans into the migration, I'm not a fan of it but the other solutions (SSDT/sqlpackage/etc.) are _significantly_ worse.

It's not a huge deal because we only use it on a handful of tables, I could see it being a huge nightmare on larger projects.

I do like the idea of being able to push the benefits of columnstore indexes (and other possible storage mechanisms provided by other servers?), not sure how abstracted you can be on that.

**Comment by tekrarCo** [comment]
We Hate SSDT/sqlpackage/etc , If we want do something we like to do it code first or in sql management studio , ef core code first is so lazy with new features of sql server, it seems if they are not even have new version of sql server and they work with so old version that nobody even use , is it possible they even dont know about Columstore Index Storage? 🤔

**Comment by tekrarCo** [comment]
> We just threw some SQL shenanigans into the migration, I'm not a fan of it but the other solutions (SSDT/sqlpackage/etc.) are _significantly_ worse.
> 
> It's not a huge deal because we only use it on a handful of tables, I could see it being a huge nightmare on larger projects.
> 
> I do like the idea of being able to push the benefits of columnstore indexes (and other possible storage mechanisms provided by other servers?), not sure how abstracted you can be on that.

Can you please share these migration codes? or if you have git/repository about it , can i see?

**Comment by StrangeWill** [comment]
It's literally just commenting out the table stand up, then using `migrationBuilder.SQL` to build the columnstore indexes, also you have to remove things like foreign keys into that table.

Very fragile, but it gets the job done.

**Comment by tekrarCo** [comment]
> It's literally just commenting out the table stand up, then using `migrationBuilder.SQL` to build the columnstore indexes, also you have to remove things like foreign keys into that table.
> 
> Very fragile, but it gets the job done.

do you know any link or website ?

**Comment by StrangeWill** [comment]
Nope, it's pretty basic if you know how to configure columnstore indexes from SQL, and if not you can find guides on that form Microsoft.

**Comment by StrangeWill** [comment]
As we look forward into more and more heavy data lifting with EF Core (its performance really allows for it), Columstore starts to become a more powerful tool to use for those tasks, any consideration in seeing this on v6?

**Comment by StrangeWill** [comment]
> We Hate SSDT/sqlpackage/etc , If we want do something we like to do it code first or in sql management studio , ef core code first is so lazy with new features of sql server, it seems if they are not even have new version of sql server and they work with so old version that nobody even use , is it possible they even dont know about Columstore Index Storage? 🤔

Came back to this thread because our team was discussing some points that we're still missing in EF Core, I just wanted to say the EF Core team is having to think of a larger ecosystem of database engines (Postgres/MySQL/etc. which we use all of them), along with the _crazy_ constraints that ColumnStore impress upon your database (lack of FKs for one, along with constraints in _how_ those indexes are built, along with mixing and matching of index types) that makes migration support non-trivial. I don't _fault_ the EF Core team for not considering this, especially considering how few teams I see in the field utilizing ColumnStore indexes (or being bit by it, not understanding it and wanting to go elsewhere).

To this day, 2 and a half years later, we're still just using our slapped-in SQL migrations... it still works for us. Still on my "nice to have" list.

**Comment by roji** [comment]
@StrangeWill thanks for the write-up. To summarize, this feature currently has a total of 11 votes, so we don't give it high priority when planning for the next feature. In addition, as you've mentioned, there's a workaround for using ColumnStore without specific EF Core support (i.e. via custom raw migrations); while that's far from perfect, it is a workaround - so the value of doing this is again less than other features where EF support is absolutely necessary.
